Add the DateTimeHelper class, add a method to IOHelperMethods.

dankwart-de opened the pull request:
Add convertMapToTwoColumnsFormattedTable to RoddyIOHelperMethods. This
was taken from one of the Roddy classes and can be used generically.

Add the DateTimeHelper class, which tries to reduces the necessity to
fiddle around with the Joda-Time and the Java time classes:
- By default, the class will parse ISO_DATE_TIME strings but can be
  constructed with different patterns and time zones.
- parseTime will use the pattern in the constructor or the default
  pattern to parse a time string.
- toDuration will convert a LocalDateTime object to a Duration object
  However, this is an approximation, as additional days of leap years
  are not taken into account.
- toDuration can also be used with values for years, days and so on
- timeSpanOf will calculate the duration between two LocalDateTimes
- isOlderThan will tell you, if a Duration is higher than a Duration
  called maxAge. Alternatively, you can also call the method with
  values for days, hours and so on.

Created a test class which covers all methods and the constructor.

Commit:
Add the DateTimeHelper class, add a method to IOHelperMethods.

Add convertMapToTwoColumnsFormattedTable to RoddyIOHelperMethods. This
was taken from one of the Roddy classes and can be used generically.

Add the DateTimeHelper class, which tries to reduces the necessity to
fiddle around with the Joda-Time and the Java time classes:
- By default, the class will parse ISO_DATE_TIME strings but can be
  constructed with different patterns and time zones.
- parseTime will use the pattern in the constructor or the default
  pattern to parse a time string.
- toDuration will convert a LocalDateTime object to a Duration object
  However, this is an approximation, as additional days of leap years
  are not taken into account.
- toDuration can also be used with values for years, days and so on
- timeSpanOf will calculate the duration between two LocalDateTimes
- isOlderThan will tell you, if a Duration is higher than a Duration
  called maxAge. Alternatively, you can also call the method with
  values for days, hours and so on.

Created a test class which covers all methods and the constructor.

vinjana commented on src/main/groovy/de/dkfz/roddy/tools/DateTimeHelper.groovy:
```diff
@@ -0,0 +1,56 @@
+        else
+            this.DATE_PATTERN = DateTimeFormatter
+                    .ofPattern(dateParserPatternPattern)
+                    .withLocale(Locale.ENGLISH)
```
Why fix the locale as ENGLISH?

dankwart-de replied:
Actually I don't know, I reused existing code. @askask , @fkaercher do you know, why you put this in?

vinjana commented on src/main/groovy/de/dkfz/roddy/tools/DateTimeHelper.groovy:
```diff
@@ -0,0 +1,56 @@
+        toDuration(higher) - toDuration(lower)
+    }
+
+    static boolean isOlderThan(Duration age, int days, int hours, int minutes, int seconds) {
```
The concept "older" does not make sense for durations. Call this "isLongerThan".

dankwart-de replied:
makes sense. I'll fix it.

vinjana commented on src/main/groovy/de/dkfz/roddy/tools/DateTimeHelper.groovy:
```diff
@@ -0,0 +1,56 @@
+        return isOlderThan(age, maxAge)
+    }
+
+    static boolean isOlderThan(Duration timeSpan, Duration maxAge) {
```
Again: A duration is not an age. ```Duration maxLength```

vinjana commented on src/main/groovy/de/dkfz/roddy/tools/DateTimeHelper.groovy:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    static Duration timeSpanOf(LocalDateTime lower, LocalDateTime higher) {
+        toDuration(higher) - toDuration(lower)
```
There is no check that one is indeed "lower" than the other. Why not call it more generically "difference" or "dateDifference" or "minus" or similar? What happens if the "higher" indeed is the "lower". Also they should be called "earlier" and "later".

dankwart-de replied:
Called it a and b. If the result is negative, I multiply it by -1 now. That works.

vinjana commented on src/main/groovy/de/dkfz/roddy/tools/DateTimeHelper.groovy:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    static Duration toDuration(LocalDateTime ldt) {
+        // There is no plusYears, we just assume 365 days for a year
```
Hm, very bad assumption. All output will be false. Also leap-seconds, etc. are ignored. This should not be called ```toDuration``` but ```toEstimatedDuration```, or a cleaner version should be implemented (which is admittedly complex and likely too complex for our purposes).

vinjana replied:
Consider that Oracle advises to use Duration when time is measured in nanoseconds, while Period when time is measured in days, hours, etc.

https://docs.oracle.com/javase/tutorial/datetime/iso/period.html

dankwart-de replied:
Thing is, I really did not figure out a better way. There are a lot of classes in java.time and you cannot use easily for what we need. It would be really nice to have something like DateTime1-DateTime2. But either I did not find it or it does not exist. That is, why I did it. I know the assumption is really bad and lacks a lot. But I also think, that for our case it should work. Our code is either using minutes (Roddy) or days (OTP), don't know how I can fix this in a good way.

vinjana replied:
What about the Period class?

dankwart-de replied:
The Period class does not allow us to use minutes or hours. It is based on days, weeks and years. I found out, that Duration has a between method which takes to LocalDateTime objects. This is a good start and I currently try to build things around it. Duration can also be negative which means you can just multiply it by -1 to get a positive duration.

vinjana commented on src/main/groovy/de/dkfz/roddy/tools/RoddyIOHelperMethods.groovy:
```diff
@@ -362,4 +362,21 @@ class RoddyIOHelperMethods {
     static String printTimingInfo(String info, long t1, long t2) {
```
Can you comment this method? It is completely unclear what info, t1 and t2 may possibly be.

vinjana commented on src/main/groovy/de/dkfz/roddy/tools/RoddyIOHelperMethods.groovy:
```diff
@@ -362,4 +362,21 @@ class RoddyIOHelperMethods {
+     * @param closureForValue
+     * @return
+     */
+    static List<String> convertMapToTwoColumnsFormattedTable(Map<String, ?> map, int cntOfTabs, String tabSep, Closure closureForValue) {
```
What I find problematic about this function is that it produces neither a clean tab-separated nor a clean space-separated and aligned table, but some kind of a mixture. Is this table supposed to be parsed?

dankwart-de replied:
No it is for console output. Not for parsing! I think there are some libraries which can do this. But this is another dependency and we use it rarely.

vinjana replied:
For console output I would just accept this. Maybe call it something like "...PrettyTable"? Something to express that this is not meant for parsing but for readability only?

dankwart-de replied:
Yes that makes sense, as we have to do a version bump, I can just rename it.

Commit:
Changes for #24 review by @vinjana

In DateTimeHelper:
- Renamed parseTime() to parseToZonedDateTime()
- Allow different locales than english in the constructor
- Removed the toDuration methods. They are not necessary
- Renamed timeSpanOf() to differenceBetween(). This method is actually
  very trivial and not too necessary. But in difference to the
  Duration.between method, this one always returns the absolute
  duration which cannot be negative.
- Renamed isOlderThan() to durationExceeds()

In DateTimeHelperTest:
- Removed tests for toDuration
- Fixed and extended ParseTime()
- Renamed and extended TimeSpanCalc() to "Test differenceBetween"()
- Renamed and extended IsOlderThan() to "Test durationExceeds"()

In RoddyIOHelperMethods:
- Renamed convertMapToTwoColumnsFormattedTable() to
  convertMapToTwoColumnsPrettyTable()
- Added comments for printTimingInfo() and
  convertMapToTwoColumnsPrettyTable()

vinjana commented on src/main/groovy/de/dkfz/roddy/tools/DateTimeHelper.groovy:
```diff
@@ -0,0 +1,45 @@
+    private final DateTimeFormatter DATE_PATTERN
+
+    DateTimeHelper(String dateParserPatternPattern = null, Locale locale = Locale.default, ZoneId timeZoneId = ZoneId.systemDefault()) {
+        if (!dateParserPatternPattern)
```
dateParserPatternPattern?

Commit:
Fixed a typo

dankwart-de committed:
Increase testibility, fix tests

* Add some test options to build.gradle, this will also show detailed error messages and exceptions in the Travis logs.
* Fix ParseTime to use the correct expectedResult for the [Europe/Berlin]
test case.
